File: filter.py
import json
from datetime import date

class song_filter:
    def __init__(self):
        self.minScore = 0
        self.maxScore = 5

        self.minDate = "2003/01/01"
        self.maxDate = date.today().strftime("%Y/%m/%d")

        self.genres = ["Ambient","Bluegrass","Blues","Brit Pop","Classical",
                       "Classical Rock","Country","Dance","Drum N Bass",
                       "Dubstep","Funk","General Rock","Goth","Grunge",
                       "Heavy Metal","Hip Hop - Modern","Hip Hop - Olskool",
                       "House","Indie","Industrial","Jazz","Latin",
                       "Miscellaneous","New Wave","Pop","Punk","R&amp;B","Ska",
                       "Techno","Trance","Video Game","Voice Acting"]

        self.allowAll = True

    def load_from_json(self, js_filter):
        js = json.loads(js_filter)

        self.minScore = js["minScore"]
        self.maxScore = js["maxScore"]

        self.minDate = js["minDate"].replace('-', '/')
        self.maxDate = js["maxDate"].replace('-', '/')

        self.genres = []
        for genre in js["genres"]:
            if genre == "R%26B":
                #Compatibility with my db.
                self.genres.append("R&amp;B")
            else:
                self.genres.append( genre )
            
        self.allowAll = (self.minScore == 0 and self.maxScore == 5)\
                        and (self.minDate == "2003/01/01" and self.maxDate == date.today().strftime("%Y/%m/%d") )\
                        and (len(self.genres) == 32)
        
    def get_sql_condition(self):
        local = ""
        if not self.allowAll:
            local = "WHERE "
            conditions = []
            #Is date condition necessary ?
            if(self.minDate != "2003/01/01" or self.maxDate != date.today().strftime("%Y/%m/%d")):
                conditions.append( """(date >= "%s" and date <= "%s")""" % (self.minDate, self.maxDate) )

            #Is Score condition necessary ?
            if( self.minScore != 0 or self.maxScore != 5 ):
                conditions.append( """(score >= %s and score <= %s)"""   % (self.minScore, self.maxScore) )

            #Is genre condition necessary ?
            if( len(self.genres) != 32 ):
                if len(self.genres) != 1:
                    conditions.append( """genre IN %s""" % str( tuple(self.genres) ) )
                else:
                    conditions.append( """genre=\"%s\"""" % self.genres[0] )

            #Create condition string
            for i in range(0, len(conditions)):
                local += conditions[i]
                #Remember te put AND between them.
                if i != len(conditions) -1:
                    local += " AND "

        return local

if __name__ == "__main__":

    TEST_JSON_FILTER_ALLOWALL="""\
    {"maxScore":5,"minDate":"2003-01-01","genres":["Ambient","Bluegrass","Blues","Brit Pop","Classical","Classical Rock","Country","Dance","Drum N Bass","Dubstep","Funk","General Rock","Goth","Grunge","Heavy Metal","Hip Hop - Modern","Hip Hop - Olskool","House","Indie","Industrial","Jazz","Latin","Miscellaneous","New Wave","Pop","Punk","R%%26B","Ska","Techno","Trance","Video Game","Voice Acting"],"minScore":0,"maxDate":"%s"}\
    """ % date.today().strftime("%Y/%m/%d")

    TEST_JSON_FILTER_SCORE="""\
    {"maxScore":5,"minDate":"2003-01-01","genres":["Ambient","Bluegrass","Blues","Brit Pop","Classical","Classical Rock","Country","Dance","Drum N Bass","Dubstep","Funk","General Rock","Goth","Grunge","Heavy Metal","Hip Hop - Modern","Hip Hop - Olskool","House","Indie","Industrial","Jazz","Latin","Miscellaneous","New Wave","Pop","Punk","R%%26B","Ska","Techno","Trance","Video Game","Voice Acting"],"minScore":4,"maxDate":"%s"}\
    """ % date.today().strftime("%Y/%m/%d")

    TEST_JSON_FILTER_GENRE="""\
    {"maxScore":5,"minDate":"2003-01-01","genres":["R%%26B","Ska","Techno","Trance","Video Game","Voice Acting"],"minScore":0,"maxDate":"%s"}\
    """ % date.today().strftime("%Y/%m/%d")
    
    obj = song_filter()
    obj.load_from_json( TEST_JSON_FILTER_GENRE )

    if obj.allowAll :
        print( "Allowing all, no where condition." )
    else:
        print( obj.get_sql_condition() )

File: src/filter.cpp
#include "filter.hpp"

static std::vector<enum genre> default_allowed_genre = {
    genre::CLASSICAL, genre::JAZZ, genre::SOLO_INSTRUMENT,

    genre::AMBIENT, genre::CHIPSTEP, genre::DANCE, genre::DRUM_N_BASS, genre::DUBSTEP,
    genre::HOUSE, genre::INDUSTRIAL, genre::NEW_WAVE, genre::SYNTHWAVE, genre::TECHNO,
    genre::TRANCE, genre::VIDEO_GAME,

    genre::HIP_HOP_MODERN, genre::HIP_HOP_OLSKOOL, genre::NERDCORE, genre::R_B,

    genre::BRIT_POP, genre::CLASSICAL_ROCK, genre::GENERAL_ROCK, genre::GRUNGE,
    genre::HEAVY_METAL, genre::INDIE, genre::POP, genre::PUNK,

    genre::CINEMATIC, genre::EXPERIMENTAL, genre::FUNK, genre::FUSION, genre::GOTH, genre::MISCELLANEOUS,
    genre::SKA, genre::WORLD,

    genre::DISCUSSION, genre::MUSIC, genre::STORYTELLING,

    genre::BLUEGRASS, genre::BLUES, genre::COUNTRY,

    genre::A_CAPELLA, genre::COMEDY, genre::CREEPYPASTA, genre::DRAMA,
    genre::INFORMATIONAL, genre::SPOKEN_WORLD, genre::VOICE_DEMO
};

Json::Value readJSONFromRequest(const Pistache::Rest::Request &req) {
    Json::CharReaderBuilder builder; Json::Value root;
    Json::CharReader *j = builder.newCharReader();
    const char *body = req.body().c_str();
    std::string err;

    j->parse(body, body + req.body().length(), &root, &err);
    delete j;
    return root;
}

#ifdef NG_AUDIO_BRIDGE_COMPAT
#include <curl/curl.h>

AudioBridgeFilter::AudioBridgeFilter() :
    m_minscore(0), m_maxscore(5),
    m_mindate("2003/01/01"),
    m_allowedgenre(default_allowed_genre)
{}

AudioBridgeFilter::~AudioBridgeFilter() {}

void AudioBridgeFilter::set(const Pistache::Rest::Request &req) {
    Json::Reader reader; Json::Value root;
    CURL *curl = curl_easy_init(); int dlen;
    auto arg = req.body().substr(11);
    if (arg.empty()) return;

    char *jsonstr = curl_easy_unescape(curl, arg.c_str(), arg.length(), &dlen);
    if (!reader.parse(jsonstr, jsonstr + dlen, root))
        goto end;

    m_minscore = root.get("minScore", 0).asFloat();
    m_maxscore = root.get("maxScore", 5).asFloat();

    std::string::size_type pos;
    m_mindate = root.get("minDate", "2003/01/01").asString();
    while ((pos = m_mindate.find('-')) != std::string::npos)
        m_mindate.replace(pos, 1, "/");

    m_maxdate = root.get("maxDate", "").asString();
    while ((pos = m_maxdate.find('-')) != std::string::npos)
        m_maxdate.replace(pos, 1, "/");

    if (root["genres"].isArray()) {
        m_allowedgenre.clear();
        for (auto i : root["genres"])
            m_allowedgenre.push_back(strToGenre(i.asString()));
    }
end:
    curl_free(jsonstr);
    curl_easy_cleanup(curl);
}

bool AudioBridgeFilter::validate() const noexcept {
    if (!std::regex_match(m_mindate, regdate)) return false;

    if (!m_maxdate.empty())
        if (!std::regex_match(m_maxdate, regdate)) return false;

    return true;
}

const std::string AudioBridgeFilter::getQuery() const noexcept {
    char *buff;
    std::ostringstream sql_query; std::vector<std::string> conditions;
    sql_query << "SELECT id,title,composer,score,genre,submission_date,url "
              << "FROM Tracks WHERE id IN (SELECT id FROM Tracks";

    if (!m_maxdate.empty() && m_mindate != "2003/01/01") {
        buff = sqlite3_mprintf("(submission_date BETWEEN %Q AND %Q)", m_mindate.c_str(), m_maxdate.c_str());
        conditions.push_back(buff);
        sqlite3_free(buff);
    }

    if (m_minscore != 0 || m_maxscore != 5)
        conditions.push_back("(score BETWEEN " + std::to_string(m_minscore) +
                             " AND " + std::to_string(m_maxscore) + ")");

    if (m_allowedgenre.size() < 48) {
        if (m_allowedgenre.size() == 1) {
            conditions.push_back("genre=" + std::to_string(static_cast<int>(m_allowedgenre[0])));
        } else {
            std::string tmp = "genre IN (";
            for (auto i = m_allowedgenre.begin(); i != m_allowedgenre.end(); i++) {
                tmp += std::to_string(static_cast<int>(*i));
                if (i != m_allowedgenre.end() - 1) tmp += ", ";
            }
            tmp += ')';
            conditions.push_back(tmp);
        }
    }

    if (!conditions.empty()) sql_query << " WHERE ";
    for (auto i = conditions.begin(); i != conditions.end(); i++) {
        sql_query << *i;
        if (i != conditions.end() - 1) sql_query << " AND ";
    }

    sql_query << " ORDER BY RANDOM() LIMIT 25);";

    return sql_query.str();
}
#endif  // NG_AUDIO_BRIDGE_COMPAT

APIFilter::APIFilter() :
    m_mindate("2003/01/01"), m_maxdate(""),
    m_minscore(0), m_maxscore(5),
    m_allowUnrated(true) {
    m_allowedgenre = default_allowed_genre;
}

APIFilter::~APIFilter() {}

void APIFilter::set(const Pistache::Rest::Request &req) {
    Json::Value root = readJSONFromRequest(req);

    m_mindate = root.get("minDate", "2003/01/01").asString();
    m_maxdate = root.get("maxDate", "").asString();
    m_minscore = root.get("minScore", 0).asFloat();
    m_maxscore = root.get("maxScore", 5).asFloat();
    m_allowUnrated = root.get("allowUnrated", true).asBool();

    if (!root["genres"].isArray()) return;

    m_allowedgenre.clear();
    for (auto i : root["genres"]) {
        if (!i.isInt()) continue;

        try {
            m_allowedgenre.push_back(static_cast<enum genre>(i.asInt()));
        } catch (std::exception&) {continue;}
    }
}

bool APIFilter::validate() const noexcept {
    // Date check.
    if (!std::regex_match(m_mindate, regdate)) return false;

    if (!m_maxdate.empty())
        if (!std::regex_match(m_maxdate, regdate)) return false;

    if (m_minscore > m_maxscore) return false;

    if (!(m_minscore >= 0 && m_minscore <= 5) || !( m_maxscore >= 0 && m_maxscore <= 5))
        return false;

    return true;
}

const std::string APIFilter::getQuery() const noexcept {
    char *buff;
    std::ostringstream sql_query; std::vector<std::string> conditions;
    sql_query << "SELECT id,title,composer,score,genre,submission_date,url "
              << "FROM Tracks WHERE id IN (SELECT id FROM Tracks";

    if (m_minscore != 0 || m_maxscore != 5) {
        std::string tmp = "(score BETWEEN " + std::to_string(m_minscore) +
                          " AND " + std::to_string(m_maxscore) + ")";

        if (m_allowUnrated)
            tmp = "(" + tmp + " OR score=-1)";

        conditions.push_back(tmp);
    }

    if (m_mindate != "2003/01/01" || !m_maxdate.empty()) {
        buff = sqlite3_mprintf("(submission_date BETWEEN %Q AND %Q)", m_mindate.c_str(), m_maxdate.c_str());
        conditions.push_back(buff);
        sqlite3_free(buff);
    }

    if (m_allowedgenre.size() < 48) {
        std::string tmp = "genre IN (";
        for (auto i = m_allowedgenre.begin(); i != m_allowedgenre.end(); i++) {
            tmp += std::to_string(static_cast<int>(*i));
            if (i != m_allowedgenre.end() - 1) tmp += ", ";
        }
        tmp += ')';
        conditions.push_back(tmp);
    }

    if (!conditions.empty()) sql_query << " WHERE ";
    for (auto i = conditions.begin(); i != conditions.end(); i++) {
        sql_query << *i;
        if (i != conditions.end() - 1) sql_query << " AND ";
    }
    sql_query << " ORDER BY RANDOM() LIMIT 25);";

    return sql_query.str();
}

APIFilterComposer::APIFilterComposer() {}

APIFilterComposer::~APIFilterComposer() {}

void APIFilterComposer::set(const Pistache::Rest::Request &req) {
    m_composer = readJSONFromRequest(req).get("composer", "").asString();
}

bool APIFilterComposer::validate() const noexcept {
    return !m_composer.empty();
}

const std::string APIFilterComposer::getQuery() const noexcept {
    std::string result; char *buff;
    buff = sqlite3_mprintf("SELECT id,title,composer,score,genre,submission_date,url "
                           "FROM Tracks WHERE composer LIKE %Q;", m_composer.c_str());
    result = buff;
    sqlite3_free(buff);
    return result;
}

void APIFilterRange::set(const Pistache::Rest::Request &req) {
    auto root = readJSONFromRequest(req);
    auto mul  = root.get("mul", 1).asInt();
    auto page = root.get("page", 1).asInt();

    m_min = page*mul + 1;
    m_max = (page+1)*mul;
}
